modules/src/timestamp.rs

```diff
@@ -0,0 +1,179 @@
+use std::convert::TryInto;
+use std::fmt::Display;
+use std::num::{ParseIntError, TryFromIntError};
+use std::str::FromStr;
+use thiserror::Error;
+
+use chrono::{offset::Utc, DateTime, TimeZone};
+
+/// A newtype wrapper over `Option<DateTime<Utc>>` to keep track of
+/// IBC packet timeout.
+///
+/// We use an explicit `Option` type to distinguish this when converting between
+/// a `u64` value and a raw timestamp. In protocol buffer, the timestamp is
+/// represented as a `u64` Unix timestamp in nanoseconds, with 0 representing the absence
+/// of timestamp.
+#[derive(PartialEq, Eq, Copy, Clone, Debug, Deserialize, Serialize, Hash)]
+pub struct Timestamp {
+    time: Option<DateTime<Utc>>,
+}
+
+/// The expiry result when comparing two timestamps.
+/// - If either timestamp is invalid (0), the result is `InvalidTimestamp`.
+/// - If the left timestamp is strictly after the right timestamp, the result is `Expired`.
+/// - Otherwise, the result is `NotExpired`.
+///
+/// User of this result may want to determine whether error should be raised,
+/// when either of the timestamp being compared is invalid.
+#[derive(PartialEq, Eq, Copy, Clone, Debug, Deserialize, Serialize, Hash)]
+pub enum Expiry {
+    Expired,
+    NotExpired,
+    InvalidTimestamp,
+}
+
+impl Timestamp {
+    /// When used in IBC, all raw timestamps are represented as u64 Unix timestamp in nanoseconds.
+    ///
+    /// A value is 0 indicates that the timestamp is not set, and result in the underlying
+    /// type being None.
+    ///
+    /// The underlying library [`chrono::DateTime`] allows conversion from nanoseconds only
+    /// from an `i64` value. In practice, `i64` still have sufficient precision for our purpose.
+    /// However we have to handle the case of `u64` overflowing in `i64`, to prevent
+    /// malicious packets from crashing the relayer.
+    pub fn from_nanoseconds(nanoseconds: u64) -> Result<Timestamp, TryFromIntError> {
+        if nanoseconds == 0 {
+            Ok(Timestamp { time: None })
+        } else {
+            let nanoseconds = nanoseconds.try_into()?;
+            Ok(Timestamp {
+                time: Some(Utc.timestamp_nanos(nanoseconds)),
+            })
+        }
+    }
+
+    /// Convert a `Timestamp` from [`chrono::DateTime<Utc>`].
+    pub fn from_datetime(time: DateTime<Utc>) -> Timestamp {
+        Timestamp { time: Some(time) }
+    }
+
+    /// Convert a `Timestamp` to `u64` value in nanoseconds. If no timestamp
+    /// is set, the result is 0.
+    pub fn as_nanoseconds(&self) -> u64 {
+        self.time
+            .map_or(0, |time| time.timestamp_nanos().try_into().unwrap())
+    }
+
+    /// Convert a `Timestamp` to an optional [`chrono::DateTime<Utc>`]
+    pub fn as_datetime(&self) -> Option<DateTime<Utc>> {
+        self.time
+    }
+
+    /// A timestamp is only valid if it is set, i.e. have non-zero raw value.
+    pub fn is_valid(&self) -> bool {
+        self.time.is_some()
+    }
```

Comment:
Should this be removed? Don't think it is used and also `None` is valid, e.g. for packets that are not timeout using the timestamp